File: fronter/src/model/quantificationLabel.ts
// 定量化ラベルのデータモデル

// ラベル付けされた定量化データ
export interface QuantificationLabel {
  id: string;

  // 言語情報
  linguistic: {
    originalText: string;              // 元のテキスト（例: "コップ半分"）
    normalizedText: string;            // 正規化テキスト
    category: string;                  // カテゴリ（量、サイズ、程度など）
    context: string;                   // 使用文脈
    domain: string;                    // ドメイン（料理、建築、デザインなど）
  };
  
  // 画像情報
  visual: {
    imageUrl: string;                  // 画像URL
    thumbnailUrl: string;              // サムネイルURL
    imageDescription: string;          // 画像の説明テキスト
    annotations: ImageAnnotation[];    // 画像アノテーション
    metadata: {
      width: number;
      height: number;
      format: string;
      capturedAt: string;
      device?: string;
    };
  };
  
  // 定量化情報
  quantification: {
    value: number;                     // 数値
    unit: string;                      // 単位
    range: {
      min: number;
      max: number;
      typical: number;                 // 典型値
      distribution: 'normal' | 'uniform' | 'skewed';
    };
    precision: number;                 // 精度（小数点以下の桁数）
    confidence: number;                // 信頼度
  };
  
  // 概念情報
  concept: {
    abstractLevel: 'concrete' | 'semi-abstract' | 'abstract';  // 抽象度
    culturalContext?: string;         // 文化的文脈
    temporalContext?: string;         // 時間的文脈
    spatialContext?: string;          // 空間的文脈
    relatedConcepts: string[];        // 関連概念
    semanticTags: string[];           // 意味タグ
  };
  
  // 評価情報
  evaluation: {
    accuracy: number;                  // 精度評価
    consistency: number;               // 一貫性評価
    reproducibility: number;           // 再現性評価
    usability: number;                 // 使いやすさ評価
    verificationCount: number;         // 検証回数
    lastVerified: string;             // 最終検証日時
  };
  
  // 履歴情報
  history: {
    createdAt: string;
    createdBy: string;
    updatedAt: string;
    updatedBy: string;
    version: number;
    revisions: LabelRevision[];
  };
  
  // メタデータ
  metadata: {
    source: 'manual' | 'automatic' | 'hybrid';  // ラベル付けソース
    validated: boolean;                          // 検証済みフラグ
    publicVisibility: boolean;                   // 公開フラグ
    tags: string[];                              // 汎用タグ
    notes?: string;                              // 備考
  };
}

// 画像アノテーション
export interface ImageAnnotation {
  id: string;
  type: 'region' | 'point' | 'measurement' | 'text';
  coordinates: {
    x: number;
    y: number;
    width?: number;
    height?: number;
  };
  label: string;
  value?: number;
  unit?: string;
  confidence: number;
  createdBy: string;
  createdAt: string;
}

// ラベル改訂履歴
export interface LabelRevision {
  revisionId: string;
  timestamp: string;
  userId: string;
  changes: {
    field: string;
    oldValue: any;
    newValue: any;
    reason?: string;
  }[];
  comment?: string;
}

// ラベルデータセット
export interface LabelDataset {
  id: string;
  name: string;
  description: string;
  domain: string;
  labels: QuantificationLabel[];
  
  statistics: {
    totalLabels: number;
    verifiedLabels: number;
    averageAccuracy: number;
    coverageByConcept: Record<string, number>;
    coverageByDomain: Record<string, number>;
  };
  
  quality: {
    completeness: number;              // データの完全性
    consistency: number;               // 一貫性
    diversity: number;                 // 多様性
    balance: number;                   // バランス
  };
  
  metadata: {
    createdAt: string;
    createdBy: string;
    lastUpdated: string;
    version: string;
    license?: string;
    citation?: string;
  };
}

// ラベル関係
export interface LabelRelation {
  id: string;
  sourceLabel: string;                // ソースラベルID
  targetLabel: string;                // ターゲットラベルID
  relationType: 'synonym' | 'hypernym' | 'hyponym' | 'meronym' | 'holonym' | 'similar' | 'opposite';
  strength: number;                    // 関係の強さ (0-1)
  bidirectional: boolean;             // 双方向関係かどうか
  context?: string;                   // 関係が成立する文脈
  confidence: number;
}

// ラベル検索クエリ
export interface LabelSearchQuery {
  text?: string;                      // テキスト検索
  domain?: string;                    // ドメインフィルタ
  category?: string;                  // カテゴリフィルタ
  valueRange?: {
    min: number;
    max: number;
    unit: string;
  };
  concepts?: string[];                // 概念フィルタ
  minConfidence?: number;             // 最小信頼度
  verified?: boolean;                 // 検証済みのみ
  dateRange?: {
    from: string;
    to: string;
  };
  limit?: number;
  offset?: number;
  sortBy?: 'relevance' | 'confidence' | 'date' | 'usage';
  sortOrder?: 'asc' | 'desc';
}

// ラベル作成リクエスト
export interface CreateLabelRequest {
  text: string;
  imageFile?: File;
  imageUrl?: string;
  description: string;
  value: number;
  unit: string;
  domain: string;
  category: string;
  concepts?: string[];
  tags?: string[];
}

// ラベル更新リクエスト
export interface UpdateLabelRequest {
  id: string;
  updates: Partial<{
    normalizedText: string;
    imageDescription: string;
    value: number;
    unit: string;
    confidence: number;
    concepts: string[];
    tags: string[];
    notes: string;
  }>;
  reason: string;                     // 更新理由
}

// ラベル検証リクエスト
export interface VerifyLabelRequest {
  labelId: string;
  verification: {
    accurate: boolean;
    consistency: boolean;
    reproducible: boolean;
    usable: boolean;
    feedback?: string;
    suggestedValue?: number;
    suggestedUnit?: string;
  };
  verifierId: string;
}

// ラベル統計
export interface LabelStatistics {
  totalLabels: number;
  labelsByDomain: Record<string, number>;
  labelsByCategory: Record<string, number>;
  labelsByConcept: Record<string, number>;
  
  averageMetrics: {
    confidence: number;
    accuracy: number;
    consistency: number;
    reproducibility: number;
  };
  
  distribution: {
    valueDistribution: Array<{
      range: string;
      count: number;
      percentage: number;
    }>;
    unitDistribution: Record<string, number>;
    conceptDistribution: Record<string, number>;
  };
  
  temporal: {
    createdLast24h: number;
    createdLast7d: number;
    createdLast30d: number;
    verifiedLast24h: number;
    verifiedLast7d: number;
    verifiedLast30d: number;
  };
  
  quality: {
    highConfidence: number;           // 信頼度80%以上
    mediumConfidence: number;         // 信頼度50-80%
    lowConfidence: number;            // 信頼度50%未満
    verified: number;
    unverified: number;
  };
}

File: backer/godotask/model/quantification_label.go
package model

import (
	"time"
	"gorm.io/gorm"

	"github.com/godotask/lib"
)

type JSON = lib.JSON

// QuantificationLabel - 定量化ラベルのメインモデル
type QuantificationLabel struct {
	ID  string    `gorm:"type:varchar(255);primaryKey" json:"id"`

	UserID int `json:"user_id" gorm:"index"`
	TaskID int  `json:"task_id" gorm:"index"`

	// 言語情報
	OriginalText    string `json:"original_text" gorm:"type:text"`
	NormalizedText  string `json:"normalized_text" gorm:"type:text"`
	Category        string `json:"category" gorm:"index"`
	// Context         *string `json:"context" gorm:"type:text"`
	Domain          string `json:"domain" gorm:"index"`

	// 画像情報
	// ImageURL         *string `json:"image_url" gorm:"type:text"`
	// ThumbnailURL     *string `json:"thumbnail_url" gorm:"type:text"`
	// ImageDescription *string `json:"image_description" gorm:"type:text"`
	// ImageMetadata    *JSON   `json:"image_metadata" gorm:"type:jsonb"`

	// 定量化情報
	// Value       *float64 `json:"value"`
	// Unit        *string  `json:"unit" gorm:"index"`
	// MinRange    *float64 `json:"min_range"`
	// MaxRange    *float64 `json:"max_range"`
	// TypicalValue *float64 `json:"typical_value"`
	// Precision   *int     `json:"precision"`
	// Confidence  *float64 `json:"confidence"`

	// 概念情報
	// AbstractLevel     *string `json:"abstract_level" gorm:"index"` // concrete, semi-abstract, abstract
	// CulturalContext   *string `json:"cultural_context"`
	// TemporalContext   *string `json:"temporal_context"`
	// SpatialContext    *string `json:"spatial_context"`
	// RelatedConcepts   *JSON   `json:"related_concepts" gorm:"type:jsonb"`
	// SemanticTags      *JSON   `json:"semantic_tags" gorm:"type:jsonb"`

	// 評価情報
	// Accuracy       *float64 `json:"accuracy"`
	// Consistency    *float64 `json:"consistency"`
	// Reproducibility *float64 `json:"reproducibility"`
	// Usability      *float64 `json:"usability"`
	// VerificationCount *int  `json:"verification_count"`
	// LastVerified   *time.Time `json:"last_verified"`

	// メタデータ
	// Source           *string `json:"source" gorm:"index"` // manual, automatic, hybrid
	// Validated        *bool   `json:"validated" gorm:"index;default:false"`
	// PublicVisibility *bool   `json:"public_visibility" gorm:"default:true"`
	// Tags             *JSON   `json:"tags" gorm:"type:jsonb"`
	// Notes            *string `json:"notes" gorm:"type:text"`

	// 履歴管理
	// Version   *int    `json:"version" gorm:"default:1"`
	// CreatedBy *string `json:"created_by"`
	// UpdatedBy *string `json:"updated_by"`

	// タイムスタンプ
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	// DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	// リレーション
	// Annotations []ImageAnnotation `json:"annotations" gorm:"foreignKey:LabelID"`
	// Revisions   []LabelRevision   `json:"revisions" gorm:"foreignKey:LabelID"`
}

// ImageAnnotation - 画像アノテーション
type ImageAnnotation struct {
	ID       string  `gorm:"type:varchar(255);primaryKey" json:"id"`
	LabelID  string  `json:"label_id" gorm:"index"`
	Type     string  `json:"type"` // region, point, measurement, text
	X        float64 `json:"x"`
	Y        float64 `json:"y"`
	Width    float64 `json:"width"`
	Height   float64 `json:"height"`
	Label    string  `json:"label"`
	Value    float64 `json:"value"`
	Unit     string  `json:"unit"`
	Confidence float64 `json:"confidence"`
	CreatedBy  string  `json:"created_by"`
	CreatedAt  time.Time `json:"created_at"`
}

// LabelRevision - ラベル改訂履歴
type LabelRevision struct {
	ID        string    `gorm:"type:varchar(255);primaryKey" json:"id"`
	LabelID   string    `json:"label_id" gorm:"index"`
	Version   int       `json:"version"`
	Changes   JSON      `json:"changes" gorm:"type:jsonb"`
	Comment   string    `json:"comment" gorm:"type:text"`
	UserID    string    `json:"user_id"`
	Timestamp time.Time `json:"timestamp"`
}

// LabelDataset - ラベルデータセット
type LabelDataset struct {
	ID          string `gorm:"type:varchar(255);primaryKey" json:"id"`
	Name        string `json:"name"`
	Description string `json:"description" gorm:"type:text"`
	Domain      string `json:"domain" gorm:"index"`
	
	// 統計情報
	TotalLabels    int     `json:"total_labels"`
	VerifiedLabels int     `json:"verified_labels"`
	AverageAccuracy float64 `json:"average_accuracy"`
	
	// 品質メトリクス
	Completeness float64 `json:"completeness"`
	Consistency  float64 `json:"consistency"`
	Diversity    float64 `json:"diversity"`
	Balance      float64 `json:"balance"`

	// メタデータ
	Version   string `json:"version"`
	License   string `json:"license"`
	Citation  string `json:"citation"`
	CreatedBy string `json:"created_by"`

	// タイムスタンプ
	CreatedAt time.Time      `json:"created_at"`
	UpdatedAt time.Time      `json:"updated_at"`
	DeletedAt gorm.DeletedAt `gorm:"index" json:"-"`

	// リレーション
	Labels []QuantificationLabel `json:"labels" gorm:"many2many:dataset_labels;"`
}

// LabelRelation - ラベル間の関係
type LabelRelation struct {
	ID           string  `gorm:"type:varchar(255);primaryKey" json:"id"`
	SourceID     string  `json:"source_id" gorm:"index"`
	TargetID     string  `json:"target_id" gorm:"index"`
	RelationType string  `json:"relation_type"` // synonym, hypernym, hyponym, etc.
	Strength     float64 `json:"strength"`
	Bidirectional bool   `json:"bidirectional"`
	Context      string  `json:"context"`
	Confidence   float64 `json:"confidence"`
	CreatedAt    time.Time `json:"created_at"`
}

// VisualMetaphor - 視覚的メタファー
type VisualMetaphor struct {
	ID              string  `gorm:"type:varchar(255);primaryKey" json:"id"`
	Metaphor        string  `json:"metaphor"`
	ReferenceObject string  `json:"reference_object"`
	Width           float64 `json:"width"`
	Height          float64 `json:"height"`
	Depth           float64 `json:"depth"`
	ImageURL        string  `json:"image_url"`
	MinVariability  float64 `json:"min_variability"`
	MaxVariability  float64 `json:"max_variability"`
	CreatedAt       time.Time `json:"created_at"`
	UpdatedAt       time.Time `json:"updated_at"`
}

// UserCalibration - ユーザーキャリブレーション
type UserCalibration struct {
	ID              string    `gorm:"type:varchar(255);primaryKey" json:"id"`
	UserID          uint      `json:"user_id" gorm:"index"`
	ReferenceObject string    `json:"reference_object"`
	Measurements    JSON      `json:"measurements" gorm:"type:jsonb"`
	ImageURL        string    `json:"image_url"`
	Confidence      float64   `json:"confidence"`
	CreatedAt       time.Time `json:"created_at"`
	UpdatedAt       time.Time `json:"updated_at"`
}

// MultimodalData - マルチモーダル処理データ
type MultimodalData struct {
	ID       string `gorm:"type:varchar(255);primaryKey" json:"id"`
	UserID   uint   `json:"user_id" gorm:"index"`
	TaskID   uint   `json:"task_id" gorm:"index"`

	// 言語特徴
	Text            string  `json:"text"`
	Tokens          JSON    `json:"tokens" gorm:"type:jsonb"`
	SemanticVector  JSON    `json:"semantic_vector" gorm:"type:jsonb"`
	AmbiguityScore  float64 `json:"ambiguity_score"`

	// 画像特徴
	ImageURL    string  `json:"image_url"`
	Objects     JSON    `json:"objects" gorm:"type:jsonb"`
	Measurements JSON   `json:"measurements" gorm:"type:jsonb"`
	ImageConfidence float64 `json:"image_confidence"`
	
	// 関連付け
	MappingType        string  `json:"mapping_type"` // direct, inferred, learned
	CorrelationScore   float64 `json:"correlation_score"`
	ContextRelevance   float64 `json:"context_relevance"`
	HistoricalAccuracy float64 `json:"historical_accuracy"`
	
	// 定量化結果
	Value      float64 `json:"value"`
	Unit       string  `json:"unit"`
	MinRange   float64 `json:"min_range"`
	MaxRange   float64 `json:"max_range"`
	Confidence float64 `json:"confidence"`
	
	// メタデータ
	Verified     bool   `json:"verified"`
	UserFeedback string `json:"user_feedback"` // correct, too_high, too_low, incorrect
	
	// タイムスタンプ
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// リクエスト/レスポンス構造体

// CreateLabelRequest - ラベル作成リクエスト
type CreateLabelRequest struct {
	Text        string   `json:"text" binding:"required"`
	Description string   `json:"description" binding:"required"`
	Value       float64  `json:"value" binding:"required"`
	Unit        string   `json:"unit" binding:"required"`
	Domain      string   `json:"domain" binding:"required"`
	Category    string   `json:"category" binding:"required"`
	ImageURL    string   `json:"image_url"`
	Concepts    []string `json:"concepts"`
	Tags        []string `json:"tags"`
}

// UpdateLabelRequest - ラベル更新リクエスト
type UpdateLabelRequest struct {
	ID      string                 `json:"id" binding:"required"`
	Updates map[string]interface{} `json:"updates" binding:"required"`
	Reason  string                 `json:"reason" binding:"required"`
}

// VerifyLabelRequest - ラベル検証リクエスト
type VerifyLabelRequest struct {
	LabelID      string                 `json:"label_id" binding:"required"`
	Verification map[string]interface{} `json:"verification" binding:"required"`
	VerifierID   string                 `json:"verifier_id" binding:"required"`
}

// LabelSearchQuery - ラベル検索クエリ
type LabelSearchQuery struct {
	Text         string   `json:"text"`
	Domain       string   `json:"domain"`
	Category     string   `json:"category"`
	MinValue     float64  `json:"min_value"`
	MaxValue     float64  `json:"max_value"`
	Unit         string   `json:"unit"`
	Concepts     []string `json:"concepts"`
	MinConfidence float64 `json:"min_confidence"`
	Verified     *bool    `json:"verified"`
	From         string   `json:"from"`
	To           string   `json:"to"`
	Limit        int      `json:"limit"`
	Offset       int      `json:"offset"`
	SortBy       string   `json:"sort_by"`
	SortOrder    string   `json:"sort_order"`
}

// LabelStatistics - ラベル統計
type LabelStatistics struct {
	TotalLabels         int                    `json:"total_labels"`
	LabelsByDomain      map[string]int         `json:"labels_by_domain"`
	LabelsByCategory    map[string]int         `json:"labels_by_category"`
	LabelsByConcept     map[string]int         `json:"labels_by_concept"`
	AverageMetrics      map[string]float64     `json:"average_metrics"`
	ValueDistribution   []map[string]interface{} `json:"value_distribution"`
	UnitDistribution    map[string]int         `json:"unit_distribution"`
	ConceptDistribution map[string]int         `json:"concept_distribution"`
	Temporal            map[string]int         `json:"temporal"`
	Quality             map[string]int         `json:"quality"`
}

// SuggestionRequest - 定量化提案リクエスト
type SuggestionRequest struct {
	Text     string `json:"text" binding:"required"`
	ImageURL string `json:"image_url"`
	Domain   string `json:"domain"`
}

// SuggestionResponse - 定量化提案レスポンス
type SuggestionResponse struct {
	Suggestions []struct {
		Value      float64 `json:"value"`
		Unit       string  `json:"unit"`
		Confidence float64 `json:"confidence"`
		Source     string  `json:"source"`
	} `json:"suggestions"`
}
